qcodes/instrument_drivers/tektronix/Keithley_2450.py

```diff
@@ -1,11 +1,171 @@
+    _raw_values: Optional[list] = None
+
+    def get_selected(self) -> Optional[list]:
+        return self._raw_values
+
+
+class Buffer2450(InstrumentChannel):
+    """
+    Treat the reading buffer as a submodule, similar to Sense and Source
+    """
+    default_buffer = {"defbuffer1", "defbuffer2"}
+
+    buffer_elements = {
+        "date": "DATE",
+        "measurement_formatted": "FORMatted",
+        "fractional_seconds": "FRACtional",
+        "measurement": "READing",
+        "relative_time": "RELative",
+        "seconds": "SEConds",
+        "source_value": "SOURce",
+        "source_value_formatted": "SOURFORMatted",
+        "source_value_status": "SOURSTATus",
+        "source_value_unit": "SOURUNIT",
+        "measurement_status": "STATus",
+        "time": "TIME",
+        "timestamp": "TSTamp",
+        "measurement_unit": "UNIT"
+    }
+
+    def __init__(
+            self,
+            parent: 'Keithley2450',
+            name: str,
+            size: Optional[int] = None,
+            style: str = ''
+    ) -> None:
+        super().__init__(parent, name)
+        self.buffer_name = name
+        self._size = size
+        self.style = style
+
+        if self.buffer_name not in self.default_buffer:
+            # when making a new buffer, the "size" parameter is required.
+            if size is None:
+                raise TypeError(
+                    "buffer() missing 1 required positional argument: 'size'"
+                )
+            self.write(
+                f":TRACe:MAKE '{self.buffer_name}', {self._size}, {self.style}"
+            )
+        else:
+            # when referring to default buffer, "size" parameter is not needed.
+            if size is not None:
+                self.log.warning(
+                    f"Please use method 'size()' to resize default buffer "
+                    f"{self.buffer_name} size to {self._size}."
+                )
+
+        self.add_parameter(
+            "size",
+            get_cmd=f":TRACe:POINts? '{self.buffer_name}'",
+            set_cmd=f":TRACe:POINts {{}}, '{self.buffer_name}'",
+            get_parser=int,
+            docstring="The number of readings a buffer can store."
+        )
+
+        self.add_parameter(
+            "number_of_readings",
+            get_cmd=f":TRACe:ACTual? '{self.buffer_name}'",
+            get_parser=int,
+            docstring="To get the number of readings in the reading buffer."
+        )
+
+        self.add_parameter(
+            "elements",
+            get_cmd=None,
+            set_cmd=None,
+            vals=Lists(Enum('date', 'measurement_formatted',
+                            'fractional_seconds', 'measurement',
+                            'relative_time', 'seconds', 'source_value',
+                            'source_value_formatted', 'source_value_status',
+                            'source_value_unit', 'measurement_status', 'time',
+                            'timestamp', 'measurement_unit')),
```

Comment:
Is there a way to map each element in the list? `mapping=buffer_elements ` won't work here.

Reply:
> map each element in the list
do you mean that you would like to have a `val_mapping` kind of behavior but allowing for "lists of these elements", right? if so, indeed, val_mapping wouldn't work for this but you could use `get_parser` and `set_parser` arguments to pass callables that implement the conversion, for example:
```python
def from_name_to_scpi(element_names: List[str]) -> List[str]:
    return [buffer_elements[element] for element in element_names]

def from_scpi_to_name(element_scpis: List[str]) -> List[str]:
    # assuming the inverted buffer_elements dict is already prepared :)
    return [inverted_buffer_elements_dict[element] for element in element_scpis]

# both of these could just be lambas, but it's up to you.

####################

# and then in add_parameter:
   ...,
   set_parser=from_name_to_scpi,
   get_parser=from_scpi_to_name,
   ...
    )
```

btw, in order not to repeat things, `buffer_elements.keys()` can simple be used: `vals=Lists(Enum(self.buffer_elements.keys()))`.

Reply:
`vals=Lists(Enum(self.buffer_elements.keys()))` I don't think this one would work, nor ` vals=Lists(Enum(list(self.buffer_elements.keys())))` -- `Enum` takes only individual variables, can't take a list of vars. There'll be a TypeError: TypeError: unhashable type: 'dict_keys', or
TypeError: unhashable type: 'list' for the latter.

Reply:
The `set_parser` and `get_parser` works like a charm~. Thanks!

Reply:
Yes that works.